text/0000-unsafe-fields.md

```diff
@@ -27,24 +27,120 @@ accompanying safety documentation.
+example, [`Vec::set_len`] is `unsafe`, because it directly manipulates its `Vec`'s length field,
+which carries the invariants that it is less than the capacity of the `Vec` and that all elements
+in the `Vec<T>` between 0 and `len` are valid `T`. It is critical that these invariants are upheld;
+if they are violated, invoking most of `Vec`'s other methods will induce undefined behavior.
+
+[`Vec::set_len`]: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len
+
+To help ensure such invariants are upheld, programmers may apply safety hygiene techniques to
+fields, denoting when they carry invariants and documenting why their uses satisfy their
+invariants. For example, the `zerocopy` crate maintains the policy that fields with safety
+invariants have `# Safety` documentation, and that uses of those fields occur in the lexical
+context of an `unsafe` block with a suitable `// SAFETY` comment.
+
+Unfortunately, Rust does not yet provide tooling support for field safety hygiene. Since the
+`unsafe` keyword cannot be applied to field definitions, Rust cannot enforce that
+potentially-invalidating uses of fields occur in the context of `unsafe` blocks, and Clippy cannot
+enforce that safety comments are present either at definition or use sites. This RFC is motivated
+by the benefits of closing this tooling gap.
+
+### Benefit: Improving Field Safety Hygiene
+
+The absence of tooling support for field safety hygiene makes its practice entirely a matter of
+programmer discipline, and, consequently, rare in the Rust ecosystem. Field safety invariants
+within the standard library are sparingly and inconsistently documented; for example, at the time
+of writing, `Vec`'s capacity invariant is internally documented, but its length invariant is not.
+
+The practice of using `unsafe` blocks to denote dangerous uses of fields with safety invariants is
+exceedingly rare, since Rust actively lints against the practice with the `unused_unsafe` lint.
+
+Alternatively, Rust's visibility mechanisms can be (ab)used to help enforce that dangerous uses
+occur in `unsafe` blocks, by wrapping type definitions in an enclosing `def` module that mediates
+construction and access through `unsafe` functions; e.g.:
 
-For a real-world example, consider the `Vec` type in the standard library. It has a `len` field that
-is used to store the number of elements present. Setting this field is exposed publicly in the
-`Vec::set_len` method, which has safety requirements:
+```rust
+/// Used to mediate access to `UnalignedRef`'s conceptually-unsafe fields.
+///
+/// No additional items should be placed in this module. Impl's outside of this module should
+/// construct and destruct `UnalignedRef` solely through `from_raw` and `into_raw`.
+mod def {
+    pub struct UnalignedRef<'a, T> {
+        /// # Safety
+        /// 
+        /// `ptr` is a shared reference to a valid-but-unaligned instance of `T`.
+        pub(self) unsafe ptr: *const T,
```

Comment:
nit: `pub(self)` is useless and can be omitted

Reply:
I opted for explicit visibility, here, to highlight that the trick in question is entirely about visibility.